torchsde/_core/methods/milstein.py

```diff
@@ -12,30 +12,67 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from ...settings import SDE_TYPES, NOISE_TYPES, LEVY_AREA_APPROXIMATIONS
+import math
+
+import torch
+
+from ...settings import SDE_TYPES, NOISE_TYPES, LEVY_AREA_APPROXIMATIONS, METHOD_OPTIONS as opt
 
 from .. import base_solver
 
 
-class Milstein(base_solver.BaseSDESolver):
+class BaseMilstein(base_solver.BaseSDESolver):
     strong_order = 1.0
     weak_order = 1.0
-    sde_type = SDE_TYPES.ito
     noise_types = (NOISE_TYPES.additive, NOISE_TYPES.diagonal, NOISE_TYPES.scalar)
     levy_area_approximations = LEVY_AREA_APPROXIMATIONS.all()
 
+    def v_term(self, I_k, dt):
+        pass
+
     def step(self, t0, y0, dt):
         assert dt > 0, 'Underflow in dt {}'.format(dt)
 
-        I_k = self.bm(t0, t0 + dt)
-        v = [delta_bm_ ** 2. - dt for delta_bm_ in I_k]
+        t1 = t0 + dt
+
+        I_k = self.bm(t0, t1)
+        v = self.v_term(I_k, dt)
 
         f_eval = self.sde.f(t0, y0)
-        g_prod_eval = self.sde.g_prod(t0, y0, I_k)
-        gdg_prod_eval = self.sde.gdg_prod(t0, y0, v)
+        g_eval = self.sde.g(t0, y0)
+        g_prod_eval = self.sde.prod(g_eval, I_k)
+
+        if opt.grad_free in self.options and self.options[opt.grad_free]:
```

Comment:
I'd prefer not to check this during the loop: it would be preferential to do
```
class BaseMilstein(...):
    def __init__(self, options, **kwargs):
        if opt.grad_free not in options:
            options[opt.grad_free] = False
        super(BaseMilstein, self).__init__(options=options, **kwargs)

    def step(...):
        ...
        if self.options[opt.grad_free]:
            ...
```

So that it's `__init__`'s responsibility to ensure that `options` is of the correct form, not `step`.

Reply:
Done!